Checkstyle config_loc is not generated and fails de build on Gradle 7.0

### Expected Behavior
When I run `./gradlew checkstyleMain` it should work

### Current Behavior

> * What went wrong:
> Execution failed for task ':checkstyleMain'.
> > Unable to create Root Module: config {/home/travis/build/CristianGM/config_loc-issue-reproducer/whatever/checkstyle.xml}, classpath {/home/travis/build/CristianGM/config_loc-issue-reproducer/build/classes/java/main:/home/travis/build/CristianGM/config_loc-issue-reproducer/build/resources/main}.

https://scans.gradle.com/s/nbi4hm4cjl5sy/failure#1

Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Property ${config_loc} has not been set	


### Context
We have a conventions plugin, it uses by default the checkstyle from another artifact, unless the user has his own checkstyle.xml
optionally it supports a suppresions file, like this:
`<property name="file" value="${config_loc}/checkstyle-suppressions.xml"/>`

The issue is that `config/checkstyle` folder doesn't exist, so Gradle doesn't add the `config_loc` but it was working on Gradle 6.8.3

I tried adding `default=""` but then Checkstyle fails with:
`ava.lang.StringIndexOutOfBoundsException: String index out of range: 0`

### Steps to Reproduce 
Reproducer: https://github.com/CristianGM/config_loc-issue-reproducer

In Travis I ran 2 jobs, one with 6.8.3 and another with 7.0, and both uploading scans
https://www.travis-ci.com/github/CristianGM/config_loc-issue-reproducer/builds/223050662

### Your Environment
Build scan URL: https://gradle.com/s/nbi4hm4cjl5sy

--- a/subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/checkstyle/CheckstylePluginVersionIntegrationTest.groovy
+++ b/subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/checkstyle/CheckstylePluginVersionIntegrationTest.groovy
@@ -19,9 +19,6 @@ package org.gradle.api.plugins.quality.checkstyle
 import org.gradle.integtests.fixtures.MultiVersionIntegrationSpec
 import org.gradle.integtests.fixtures.TargetCoverage
 import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
-import org.gradle.internal.reflect.problems.ValidationProblemId
-import org.gradle.internal.reflect.validation.ValidationMessageChecker
-import org.gradle.internal.reflect.validation.ValidationTestFor
 import org.gradle.quality.integtest.fixtures.CheckstyleCoverage
 import org.gradle.util.Matchers
 import org.gradle.util.Resources
@@ -37,14 +34,13 @@ import static org.hamcrest.CoreMatchers.containsString
 import static org.hamcrest.CoreMatchers.startsWith
 
 @TargetCoverage({ CheckstyleCoverage.getSupportedVersionsByJdk() })
-class CheckstylePluginVersionIntegrationTest extends MultiVersionIntegrationSpec implements ValidationMessageChecker {
+class CheckstylePluginVersionIntegrationTest extends MultiVersionIntegrationSpec {
     @Rule
     public final Resources resources = new Resources()
 
     def setup() {
         writeBuildFile()
         writeConfigFile()
-        expectReindentedValidationMessage()
     }
 
     def "analyze good code"() {
@@ -63,12 +59,8 @@ class CheckstylePluginVersionIntegrationTest extends MultiVersionIntegrationSpec
         file("build/reports/checkstyle/test.html").assertContents(containsClass("org.gradle.TestClass2"))
     }
 
-    @ValidationTestFor(
-        ValidationProblemId.INPUT_FILE_DOES_NOT_EXIST
-    )
-    def "does not support fallback when configDirectory does not exist"() {
+    def "supports fallback when configDirectory does not exist"() {
         goodCode()
-        def missing = file("config/does-not-exist")
         buildFile << """
             checkstyle {
                 config = project.resources.text.fromString('''<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
@@ -88,10 +80,26 @@ class CheckstylePluginVersionIntegrationTest extends MultiVersionIntegrationSpec
         """
 
         expect:
-        fails('check')
-        failureDescriptionContains(inputDoesNotExist {
-            type('org.gradle.api.plugins.quality.Checkstyle').property('configDirectory').dir(missing)
-        })
+        succeeds('check')
+    }
+
+    def "changes to files in config dir causes task to be out-of-date"() {
+        goodCode()
+        when:
+        succeeds('check')
+        then:
+        result.assertTaskExecuted(":checkstyleMain")
+
+        when:
+        succeeds('check')
+        then:
+        result.assertTaskSkipped(":checkstyleMain")
+
+        when:
+        file("config/checkstyle/new-file.xml").touch()
+        succeeds('check')
+        then:
+        result.assertTaskNotSkipped(":checkstyleMain")
     }
 
     @ToBeImplemented

--- a/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/Checkstyle.java
+++ b/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/Checkstyle.java
@@ -31,7 +31,7 @@ import org.gradle.api.tasks.CacheableTask;
 import org.gradle.api.tasks.Classpath;
 import org.gradle.api.tasks.Console;
 import org.gradle.api.tasks.Input;
-import org.gradle.api.tasks.InputDirectory;
+import org.gradle.api.tasks.InputFiles;
 import org.gradle.api.tasks.Internal;
 import org.gradle.api.tasks.Nested;
 import org.gradle.api.tasks.Optional;
@@ -239,9 +239,8 @@ public class Checkstyle extends SourceTask implements VerificationTask, Reportin
      * @return path to other Checkstyle configuration files
      * @since 6.0
      */
-    @Optional
     @PathSensitive(PathSensitivity.RELATIVE)
-    @InputDirectory
+    @InputFiles
     public DirectoryProperty getConfigDirectory() {
         return configDirectory;
     }

--- a/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java
+++ b/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java
@@ -59,14 +59,7 @@ public class CheckstylePlugin extends AbstractCodeQualityPlugin<Checkstyle> {
         extension = project.getExtensions().create("checkstyle", CheckstyleExtension.class, project);
         extension.setToolVersion(DEFAULT_CHECKSTYLE_VERSION);
         Directory directory = project.getRootProject().getLayout().getProjectDirectory().dir(CONFIG_DIR_NAME);
-        extension.getConfigDirectory().convention(project.provider(() -> {
-            // This provider becoming an input needs either a valid value or no value at all
-            if (directory.getAsFile().exists()) {
-                return directory;
-            } else {
-                return null;
-            }
-        }));
+        extension.getConfigDirectory().convention(directory);
         extension.setConfig(project.getResources().getText().fromFile(extension.getConfigDirectory().file("checkstyle.xml")
             // If for whatever reason the provider above cannot be resolved, go back to default location, which we know how to ignore if missing
             .orElse(directory.file("checkstyle.xml"))));

